elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/IndexDsl.scala

```diff
@@ -51,13 +51,20 @@ trait IndexDsl extends DslCommons {
     override lazy val toJsonStr = operations.map(_.toJsonStr).mkString("", "\n", "\n")
   }
 
-  case class BulkOperation(operation: OperationType, location: Option[(Index, Type)], document: Document, retryOnConflictOpt: Option[Int] = None) extends EsOperation {
+  // When upsertOpt is specified, its content is used for upsert as described in
+  // https://www.elastic.co/guide/en/elasticsearch/reference/2.3/docs-update.html
+  // When upsertOpt is not given, document is used for upsert.
+  case class BulkOperation(operation: OperationType, location: Option[(Index, Type)], document: Document, retryOnConflictOpt: Option[Int] = None, upsertOpt: Option[Document] = None) extends EsOperation {
     import EsOperation.compactJson
     override def toJson: Map[String, Any] = throw new UnsupportedOperationException
     def toJsonStr: String = {
       val (doc, retryOpt) = operation match {
         case `update` =>
-          (Document(document.id, Map("doc"->document.data) ++ Map("detect_noop" -> true, "doc_as_upsert" -> true)), retryOnConflictOpt.map(n => Map("_retry_on_conflict" -> n)))
+          if (upsertOpt.isEmpty) {
+            (Document(document.id, Map("doc"->document.data) ++ Map("detect_noop" -> true, "doc_as_upsert" -> true)), retryOnConflictOpt.map(n => Map("_retry_on_conflict" -> n)))
+          } else {
+            (Document(document.id, Map("doc"->document.data) ++ Map("detect_noop" -> true) ++ Map("upsert" -> upsertOpt.getOrElse(document).data)), retryOnConflictOpt.map(n => Map("_retry_on_conflict" -> n)))
+          }
```

Comment:
Hm, looking at this more carefully, I realize that `BulkOperation` is more generic than I had previously thought, so adding `upsertOpt` as a parameter here doesn't look quite right.

`upsertOpt` makes sense only when the `BulkOperation` is of type `update`, and not otherwise. It seems clunky to me to have an API that allows us to specify that parameter for other kinds of operations.

Reply:
That is true from API point of view. Then we will need to add an API for BulkUpdateOperation, which is not too bad I guess.

Reply:
That said, I don't know how to avoid this clunkiness while maintaining backward incompatibility.

One way to fix the issue would have been to have a different case class for each kind of operation, e.g., `BulkUpdateOperation`, `BulkDeleteOperation`, etc.

Doing that, however, would have required changes to `Bulk`s argument types.